### PR TITLE
Only print *uncaught* errors automatically 

### PR DESCRIPTION
**Pull request by roystgnr**
We're increasingly using the ability to catch and recover from some errors, and in cases where we do that it's probably better to let the catch block decide what (if anything) to print about the caught error.  By ensuring `exception::what()` is a valid description of the error, we can wait until an error actually forces us to terminate before we print that description.

Pinging @lindsayad, because although I *think* this design also obviates most of the `lock_singleton_spin_mutex()` use cases (output from `std::terminate()` should only happen after all other threads have been killed) but I'm not certain.

### CODE REVIEW

File: include/base/libmesh_common.h [hunk 1]
```diff
@@ -488,7 +405,7 @@ struct casting_compare {
 #define libmesh_exceptionless_error_msg(msg)                            \
   do {                                                                  \
     libMesh::Threads::lock_singleton_spin_mutex();                      \
-    libMesh::err << msg << std::endl;                                   \
+    libMesh::err << msg << '\n';                                        \
```
Comment (lindsayad) [inline include/base/libmesh_common.h:408]: Why is this change necessary? (Answer will apply to following occurrences)

File: src/base/libmesh.C [hunk 2]
```diff
@@ -276,6 +276,23 @@ std::terminate_handler old_terminate_handler;
 
 void libmesh_terminate_handler()
 {
+  // If we have an active exception, it may have an error message that
+  // we should print.
+  std::cout << "libMesh terminating";
```
Comment (lindsayad) [inline src/base/libmesh.C:281]: Should we regard this as `out` instead of `err`?

**Review by lindsayad — COMMENTED** [review]
I'll need to remember in what threaded context I ran into the race conditions in order to test this change

File: include/base/libmesh_common.h (hunk 1 above)
Comment (roystgnr) [reply include/base/libmesh_common.h:408]: Not necessary (we're about to die, a few redundant buffer flushes aren't going to kill anything), I guess I just noticed that at this point we're already about to call `report_error()` which is always going to call `here()` which is always going to output `std::endl` anyway, so no reason not to wait until then to flush.

File: src/base/libmesh.C (hunk 2 above)
Comment (roystgnr) [reply src/base/libmesh.C:281]: My thought had been that we don't need any repetition at this point; having multiple outputs of `what()` can be useful if an error manifests differently on different processors, but the fact that something hit the terminate handler doesn't bear repeating.

But thinking about it more... we might not have hit the terminate handler, on processor 0, we might only have thrown on other ranks.  This ought to be going to `libMesh::err` too.

**Comment by roystgnr** [comment]
I was kind of fearing those failures.  We have some MOOSE code that catches a libMesh exception, discards it (including its `what()` and rethrows a `MooseException`, and gets used in tests that expect to see the libMesh error message details afterward.  We'll need a few small changes there before this can pass.

**Comment by lindsayad** [comment]
That's mostly my code I think :grimacing:

**Comment by roystgnr** [comment]
Eh, no shame there.  This is a minor libMesh design flaw dating back a decade; no reason to be prepared for it to be fixed any minute now.

**Comment by lindsayad** [comment]
This leads to a data race in `report_error` for `reporting_error`
```c++
void report_error(const char * file, int line, const char * date, const char * time, std::ostream & os)
{
  // It is possible to have an error *inside* report_error; e.g. from
  // print_trace.  We don't want to infinitely recurse.
  static bool reporting_error = false;
  if (reporting_error)
    {
      // I heard you like error reporting, so we put an error report
      // in report_error() so you can report errors from the report.
      os << "libMesh encountered an error while attempting to report_error." << std::endl;
      return;
    }
  reporting_error = true;
```

**Comment by lindsayad** [comment]
full traces
```
WARNING: ThreadSanitizer: data race (pid=1820452)
  Read of size 1 at 0x7fc73ca13ec9 by thread T8:
    #0 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&) /home/lindad/projects4/moose/libmesh/build/../src/base/libmesh_common.C:74:7 (libmesh_devel.so.0+0x520ad2)
    #1 libMesh::DenseMatrix<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >::_lu_decompose() /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/dense_matrix_impl.h:675:7 (libmoose-devel.so.0+0x22d8d95)
    #2 libMesh::DenseMatrix<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >::lu_solve(libMesh::DenseVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> > const&, libMesh::DenseVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >&) /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/dense_matrix_impl.h:544:17 (libmoose-devel.so.0+0x22d7ec6)
    #3 INSFVVelocityVariable::adGradSln(libMesh::Elem const*, bool) const /home/lindad/projects4/moose/modules/navier_stokes/src/variables/INSFVVelocityVariable.C:367:9 (libnavier_stokes-devel.so.0+0x5792f4)
    #4 MooseVariableFV<double>::uncorrectedAdGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/framework/src/variables/MooseVariableFV.C:697:45 (libmoose-devel.so.0+0x15002b6)
    #5 INSFVVelocityVariable::uncorrectedAdGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/modules/navier_stokes/src/variables/INSFVVelocityVariable.C:99:31 (libnavier_stokes-devel.so.0+0x577f86)
    #6 MooseVariableFV<double>::adGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/framework/src/variables/MooseVariableFV.C:746:42 (libmoose-devel.so.0+0x14fe64f)
    #7 MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> Moose::FV::gradUDotNormal<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true>, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >(MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> const&, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> const&, FaceInfo const&, MooseVariableFV<double> const&, bool) /home/lindad/projects4/moose/framework/src/utils/MathFVUtils.C:44:17 (libmoose-devel.so.0+0x221f963)
    #8 FVFluxKernel::gradUDotNormal() const /home/lindad/projects4/moose/framework/src/fvkernels/FVFluxKernel.C:342:10 (libmoose-devel.so.0+0x1bb746f)
    #9 INSFVMomentumDiffusion::computeStrongResidual() /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C:49:21 (libnavier_stokes-devel.so.0+0x62ca50)
    #10 INSFVMomentumDiffusion::gatherRCData(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C:92:30 (libnavier_stokes-devel.so.0+0x62d238)
    #11 non-virtual thunk to INSFVMomentumDiffusion::gatherRCData(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C (libnavier_stokes-devel.so.0+0x62d735)
    #12 GatherRCDataFaceThread<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> >::onFace(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/build/header_symlinks/GatherRCDataFaceThread.h:109:8 (libnavier_stokes-devel.so.0+0x5e7062)
    #13 ThreadedFaceLoop<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> >::operator()(libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> const&, bool) /home/lindad/projects4/moose/framework/build/header_symlinks/ComputeFVFluxThread.h:207:9 (libnavier_stokes-devel.so.0+0x5e6abc)
    #14 void* libMesh::Threads::run_body<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*>, GatherRCDataFaceThread<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> > >(void*) /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/threads_pthread.h:236:3 (libnavier_stokes-devel.so.0+0x60078e)

  Previous write of size 1 at 0x7fc73ca13ec9 by thread T9:
    #0 libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*, std::ostream&) /home/lindad/projects4/moose/libmesh/build/../src/base/libmesh_common.C:81:19 (libmesh_devel.so.0+0x520b58)
    #1 libMesh::DenseMatrix<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >::_lu_decompose() /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/dense_matrix_impl.h:675:7 (libmoose-devel.so.0+0x22d8d95)
    #2 libMesh::DenseMatrix<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >::lu_solve(libMesh::DenseVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> > const&, libMesh::DenseVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >&) /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/dense_matrix_impl.h:544:17 (libmoose-devel.so.0+0x22d7ec6)
    #3 INSFVVelocityVariable::adGradSln(libMesh::Elem const*, bool) const /home/lindad/projects4/moose/modules/navier_stokes/src/variables/INSFVVelocityVariable.C:367:9 (libnavier_stokes-devel.so.0+0x5792f4)
    #4 MooseVariableFV<double>::uncorrectedAdGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/framework/src/variables/MooseVariableFV.C:697:45 (libmoose-devel.so.0+0x15002b6)
    #5 INSFVVelocityVariable::uncorrectedAdGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/modules/navier_stokes/src/variables/INSFVVelocityVariable.C:99:31 (libnavier_stokes-devel.so.0+0x577f86)
    #6 MooseVariableFV<double>::adGradSln(FaceInfo const&, bool) const /home/lindad/projects4/moose/framework/src/variables/MooseVariableFV.C:746:42 (libmoose-devel.so.0+0x14fe64f)
    #7 MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> Moose::FV::gradUDotNormal<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true>, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> >(MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> const&, MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> const&, FaceInfo const&, MooseVariableFV<double> const&, bool) /home/lindad/projects4/moose/framework/src/utils/MathFVUtils.C:44:17 (libmoose-devel.so.0+0x221f963)
    #8 FVFluxKernel::gradUDotNormal() const /home/lindad/projects4/moose/framework/src/fvkernels/FVFluxKernel.C:342:10 (libmoose-devel.so.0+0x1bb746f)
    #9 INSFVMomentumDiffusion::computeStrongResidual() /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C:49:21 (libnavier_stokes-devel.so.0+0x62ca50)
    #10 INSFVMomentumDiffusion::gatherRCData(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C:92:30 (libnavier_stokes-devel.so.0+0x62d238)
    #11 non-virtual thunk to INSFVMomentumDiffusion::gatherRCData(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/src/fvkernels/INSFVMomentumDiffusion.C (libnavier_stokes-devel.so.0+0x62d735)
    #12 GatherRCDataFaceThread<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> >::onFace(FaceInfo const&) /home/lindad/projects4/moose/modules/navier_stokes/build/header_symlinks/GatherRCDataFaceThread.h:109:8 (libnavier_stokes-devel.so.0+0x5e7062)
    #13 ThreadedFaceLoop<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> >::operator()(libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> const&, bool) /home/lindad/projects4/moose/framework/build/header_symlinks/ComputeFVFluxThread.h:207:9 (libnavier_stokes-devel.so.0+0x5e6abc)
    #14 void* libMesh::Threads::run_body<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*>, GatherRCDataFaceThread<libMesh::StoredRange<__gnu_cxx::__normal_iterator<FaceInfo const* const*, std::vector<FaceInfo const*, std::allocator<FaceInfo const*> > >, FaceInfo const*> > >(void*) /home/lindad/projects4/moose/scripts/../libmesh/installed/include/libmesh/threads_pthread.h:236:3 (libnavier_stokes-devel.so.0+0x60078e)
```

**Comment by roystgnr** [comment]
Nice catch.  That's a static bool there, of course it's a race condition waiting to happen.

In hindsight it feels like that was the wrong fix for its problem, too - we're in a very narrow code path and that's the best way we can come up with to keep it from recursing infinitely?  Let me see if I can come up with anything better.

**Comment by lindsayad** [comment]
It is somewhat of an arduous process:
- configure PETSc with `--with-openmp=0`
- `export CXXFLAGS=-fsanitize=thread && export LDFLAGS=-fsanitize=thread`
- configure libMesh with `--with-thread-model=pthread --disable-openmp`
- The MOOSE test to run is in `navier_stokes`: `finite_volume/ins/channel-flow.singular_two_term_expansion`

**Comment by lindsayad** [comment]
I'm happy to re-run myself if you don't feel like doing all that right now

**Comment by lindsayad** [comment]
I am pretty sure that you can do this all with gcc now as I believe all the sanitizers are being supported there, but my "thread debugging" stack is with clang so I can't say for sure

**Comment by roystgnr** [comment]
If you have time, I'd appreciate it; thanks!

File: include/base/libmesh_exceptions.h [hunk 3]
```diff
@@ -65,7 +65,8 @@ class NotImplemented : public std::logic_error
 class FileError : public std::runtime_error
 {
 public:
-  FileError(const std::string & filename) : std::runtime_error( "Error accessing file: " + filename ) {}
+  FileError(const std::string & filename, const std::string msg="") :
+    std::runtime_error("Error with file `" + filename + "'\n" + std::move(msg)) {}
```
Comment (lindsayad) [inline include/base/libmesh_exceptions.h:69]: Looks like an erroneous backtick here?

Comment (roystgnr) [reply include/base/libmesh_exceptions.h:69]: Nah, that's deliberate - see the removed code in `libmesh_file_error_msg`.  There's a "matching" neutral-tick on the other side; the intention is to try to be clearer about files/paths with spaces in them.

Comment (lindsayad) [reply include/base/libmesh_exceptions.h:69]: :+1:
